### PR TITLE
parser: Initial implementation of pg test_decoding parser

### PR DESCRIPTION
**Pull request by jonatasnona**

### CODE REVIEW

File: parser/parser_test.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	p := NewParser()
+
+	file, err := os.Open("data/transaction.txt")
+	if err != nil {
+		t.Error(err)
+	}
+
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
```
Comment (drrzmr) [inline parser/parser_test.go]: scanner could be a great feature on **tester/file**

File: parser/parser_test.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	require.Equal(t, p.s[0], line)
+	line = "COMMIT 627"
+	p.Read(line)
+	require.Equal(t, p.s[1], line)
+}
+
+func TestParserParse(t *testing.T) {
+	p := NewParser()
+
+	file, err := os.Open("data/transaction.txt")
```
Comment (drrzmr) [inline parser/parser_test.go]: **testdata**

File: parser/parser.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	p.s = append(p.s, str)
+}
+
+// Parse parse log lines
+func (p *Parser) Parse() Transaction {
+
+	t := Transaction{}
+
+	for _, line := range p.s {
+		if begin.MatchString(line) && p.fsm.Current() == "idle" {
```
Comment (drrzmr) [inline parser/parser.go]:
```
switch p.fsm.Current() {
case "idle":

}
```

File: parser/parser_test.go (hunk 1 above)
Comment (jonatasnona) [reply parser/parser_test.go]: @drrzmr could you please explain more about this or show me some example?
i do not quite understanding what you mean :confused:

Comment (drrzmr) [reply parser/parser_test.go]: Look this: https://github.com/pagarme/warp-pipe/pull/15/commits/53a7e47ae2ede9def7ec1183aa3b3016d710e73a

We could implement a test helper to wrap file load and scan, so in test we could remove this:

```
f, err := os.Open("testdata/transaction.txt")
	if err != nil {
		t.Error(err)
	}

	defer f.Close()

	scanner := bufio.NewScanner(f)

	for scanner.Scan() {
		p.Read(scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		t.Error(err)
	}
}
```

And use something like this on test:

```
err := file.ForLine(t, "transaction", func(line string) {

	p.Read(line)

})

require.NoError(err)
```

ForLine implementation could be something like this:

```
type ForLineFunc func(line string)

func ForLine(t *testing.T, filename string, cb ForLineFunc) {
	t.Helper()

	fn := fmt.Sprintf("%s.%s", filename, Config.InputExtension)
	path := filepath.Join("testdata", fn)
	
	f, err := os.Open(fn)
	require.NoError(t, err)
	defer f.Close()

	scanner := bufio.NewScanner(f)

	for scanner.Scan() {
		cb(scanner.Text())
	}
	
	err := scanner.Err()
	requite.NoError(t, err)
}
```

**Comment by drrzmr** [comment]
Looks good to me, just a minor improvement, I make this here: https://github.com/drrzmr/warp-pipe/tree/parser/test_decoding, all conflicts are fixed too. I'm approving it, but, please reset to my branch before merge this PR

**Pull request merged**